test/jdk/java/net/httpclient/DebugLoggerTest.java

```diff
@@ -0,0 +1,321 @@
+         */
+        public void stopRecording() {
+            flush(); // make sure everything that was printed before is flushed
+            synchronized (recordedData) {
+                recording = false;
+            }
+        }
+
+        /**
+         * Flushes the stream, drains the recorded data, convert it
+         * to a string, and returns it. This has the effect of
+         * flushing any recorded data from memory: the next call
+         * to {@code drainRecordedData()} will not return any data
+         * previously returned.
+         * This method can be called regardless of whether recording
+         * is started or stopped.
+         */
+        public String drainRecordedData() {
+            flush(); // make sure everything that was printed before is flushed
+            synchronized (recordedData) {
+                String data = recordedData.toString(charset);
+                recordedData.reset();
+                return data;
+            }
+        }
+
+        @Override
+        public void write(int b) {
+            super.write(b);
+            if (recording) {
+                synchronized (recordedData) {
+                    if (recording) recordedData.write(b);
+                }
+            }
+        }
+
+        @Override
+        public void write(byte[] buf, int off, int len) {
+            super.write(buf, off, len);
+            if (recording) {
+                synchronized (recordedData) {
+                    if (recording) recordedData.write(buf, off, len);
+                }
+            }
+        }
+    }
+
+    static class TestHandler extends Handler {
+        final CopyOnWriteArrayList logs = new CopyOnWriteArrayList();
+        TestHandler() {
+            setLevel(Level.ALL);
+        }
+        @Override
+        public void publish(LogRecord record) {
+            logs.add(record);
+        }
+        @Override
+        public void flush() {
+        }
+        @Override
+        public void close() throws SecurityException {
+        }
+    }
+
+    enum Destination {OUT, ERR, LOG}
+
+    static Set<Destination> getDestinations(String prop) {
+        if (prop == null) return Set.of();
+        String[] values = prop.split(",");
+        if (values.length == 0) return Set.of();
+        Set<Destination> dest = new HashSet<>(3);
+        for (var v : values) {
+            v = v.trim().toLowerCase(Locale.ROOT);
+            if ("log".equals(v)) dest.add(Destination.LOG);
+            if ("err".equals(v)) dest.add(Destination.ERR);
+            if ("out".equals(v)) dest.add(Destination.OUT);
+            if ("true".equals(v)) dest.addAll(EnumSet.of(Destination.ERR, Destination.LOG));
+        }
+        return Set.copyOf(dest);
+    }
+
+    public static void main(String[] args) {
+        err =  new RecordingPrintStream(stdErr, true, UTF_8);
+        out =  new RecordingPrintStream(stdOut, true, UTF_8);
+        System.setErr(err);
+        System.setOut(out);
```

Comment:
Should we reset these back to their original streams in a `finally` block?

Reply:
The test runs in /othervm and the RecordingPrintStream acts like a `tee` so it should not matter.